### PR TITLE
[Python Lab] More robust progress reporting

### PR DESCRIPTION
**Pull request by molly-moen**
This PR encapsulates a few items around progress reporting in Python Lab:
- Start using `onContinueOrFinish`, a new lab2 thunk that handles going to the next level/redirecting to finish. It also reports a success progress state if the level does not have conditions. I updated it in 2 ways:
   - Only send success report if the level does not have conditions, and is not a submittable or predict levels, as those levels handle progress differently
   - Wait for the success report to complete before navigating to the next level, as the next level/finish url could cause a page reload
- Have the validated instructions status icon match the progress status in the header bubbles. Previously it would always start out as a green circle, even if the level was already "passed" when the user loaded the page. It also would turn green for levels without validation as soon as the user ran their code. 
- Mark a level as in-progress after the user either edits their code or runs their code
- For a level without validation (that is not a predict level), only show the continue button once a user has edited and run their code. Predict levels still show the continue button just after run, since you can't edit a predict level.

We discussed the definition of "edit" in [this slack thread](https://codedotorg.slack.com/archives/C06JELCAPT9/p1729007699382559). We decided to go with the most permissive version of editing, which includes any file management/tab management as well as file edits. 

## Screen recording

https://github.com/user-attachments/assets/1a583c1a-6cf3-4829-972b-916298f945b5

## Links

- jira ticket: [CT-871](https://codedotorg.atlassian.net/browse/CT-871)

## Testing story
Tested locally, and verified my changes do not impact Music Lab and AI Chat, which also use the helper thunk.

## Follow-up work
This PR causes an intended regression. We previously had hard-coded that the finish button will take you to the script overview page. However, there is a backend setting to determine what a unit should do on finish, which we use in `onContinueOrFinish`. This setting is not currently on for the data science script. In order to turn it on curriculum needs to set the initiative to Foundations of CS or Foundations of Programming, and we need to complete [CT-878](https://codedotorg.atlassian.net/browse/CT-878) to make those initiatives use this setting.

I also noticed while working on this that when the progress state is updated for a bubble choice level, the header bubble is not immediately updating. I believe this was a pre-existing issue that is more obvious now because of the in-progress state. I also suspect this impacts all of lab2. I filed [CT-877](https://codedotorg.atlassian.net/browse/CT-877) to address this.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/code-studio/progressRedux.ts [hunk 1]
```diff
@@ -354,9 +360,9 @@ export function navigateToNextLevel(): ProgressThunkAction {
 
 // The user has successfully completed the level and the page
 // will not be reloading. Currently only used by Lab2 labs.
-export function sendSuccessReport(appType: string): ProgressThunkAction {
+export function sendSuccessReport(appType: string): AsyncProgressThunkAction {
```
Comment (molly-moen) [inline apps/src/code-studio/progressRedux.ts:363]: I updated these to async so we could await their completion before navigating

File: apps/src/lab2/views/components/versionHistory/VersionHistoryDropdown.tsx [hunk 2]
```diff
@@ -95,13 +93,6 @@ const VersionHistoryDropdown: React.FunctionComponent<
     }
   }, [versionList, selectedVersion, latestVersion]);
 
-  const resetVersionState = useCallback(() => {
```
Comment (molly-moen) [inline apps/src/lab2/views/components/versionHistory/VersionHistoryDropdown.tsx:98]: I moved this logic to `ProjectContainer`, which was a more logical place for it now that we also have a `hasEdited` state to reset.

File: apps/src/lab2/progress/continueOrFinishLesson.ts [hunk 3]
```diff
@@ -1,14 +1,14 @@
-import {AnyAction, ThunkAction} from '@reduxjs/toolkit';
+import {AnyAction, ThunkAction, ThunkDispatch} from '@reduxjs/toolkit';
```
Comment (molly-moen) [inline apps/src/lab2/progress/continueOrFinishLesson.ts:1]: @sanchitmalhotra126 tagging you since I refactored this function a bit.

**Review by thomasoniii — APPROVED** [review]
lgtm 🚀

**Review by sanchitmalhotra126 — APPROVED** [review]
LGTM!

**Pull request merged**